### PR TITLE
add `conftest.py`

### PR DESCRIPTION
**Pull request by davidegraff**
## Description
This PR focuses on code quality improvement in our testing suite through the use of a `conftest.py`

## Current workflow
We currently treat many test modules independently when they share the same underlying abstract data, e.g., many use dummy SMILES strings or target values. These test modules each have their own source of dummy data (through `@pytest.fixture`s), but there's no reason to store this independently. Not only does it lead to copied code across test modules, but it also slows our testing down (slightly) because we repeatedly perform the same computations (e.g. parsing the same SMILES string into a molecule across many tests)

## New workflow
The pytest `conftest.py` allows us to abstract out these details into a "central" source of `@pytest.fixture`s that can be shared among all tests in a given module. This allows us to avoid copying code, cleaning up the testing sources, and improving CI speeds.

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

File: tests/unit/data/test_datapoint.py [hunk 1]
```diff
@@ -58,15 +34,14 @@ def features_generators():
...
     with pytest.raises(ValueError):
-        MoleculeDatapoint.from_smi(smi, y=targets, x_f=features, mfs=features_generators)
+        MoleculeDatapoint.from_smi(SMI, y=targets, x_f=features, mfs=features_generators)
 
 
 def test_num_tasks(targets):
-    d = MoleculeDatapoint.from_smi("c1ccccc1", y=targets)
+    d = MoleculeDatapoint.from_smi(SMI, y=targets)
 
     assert d.t == targets.shape[0]
```
Comment (KnathanM) [inline tests/unit/data/test_datapoint.py:46]: `targets` is set such that this test runs with 1 task and 2 tasks, which is wanted. The next test, `test_addh` is run twice for each SMILES because `target` is still set to have two values. I think `targets` could be better defined in `test_addh` and `test_replace_token` as is done in `test_features_and_fg` so they are run half as many times. Perhaps that is outside the scope of your PR.

File: tests/data/test_smiles.csv
Comment (KnathanM) [inline tests/data/test_smiles.csv:1]: I agree with your rename and want to raise the question, do we need this file? `data/test_smiles.csv` has 10 SMILES. `data/smis.csv` has 100 SMILES. Do we have a use case to keep both? If we do want to keep this, maybe we want to keep the associated `data/regression_test_smiles_preds.csv` file.

File: tests/data/regression_multimolecule.csv
Comment (KnathanM) [inline tests/data/regression_multimolecule.csv:1]: The first 100 entries in this file are the same as `data/regression/mol+mol.csv`. I assume we keep the shorter one because testing 100 is faster than testing 500 without loss of test accuracy. Before we throw away this file though, is this extra data saved elsewhere? @kevingreenman

File: tests/data/regression_test_smiles_preds.csv
Comment (KnathanM) [inline tests/data/regression_test_smiles_preds.csv:1]: @kevingreenman I see you added this file two week ago. Are we okay to remove it?

File: tests/data/reaction_regression.csv
Comment (KnathanM) [inline tests/data/reaction_regression.csv:1]: Similar comment as `tests/data/regression_multimolecule.csv`. This file is a longer version of `data/regression/rxn.csv`. @kevingreenman

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@pytest.fixture
+def mol_data(mols, targets):
+    return [MoleculeDatapoint(mol, y) for mol, y in zip(mols, targets)]
+
+
+@pytest.fixture
+def rxn_data(rxns, targets):
+    return [MoleculeDatapoint(mol, y) for mol, y in zip(mols, targets)]
```
Comment (KnathanM) [inline tests/conftest.py:43]: Should this be `ReactionDatapoint` acting on two mols?

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    return request.param
+
+
+@pytest.fixture(params=_DF.mol.sample(5))
+def mol(request):
+    return request.param
+
+
+@pytest.fixture
+def mol_regression_data(data_dir):
```
Comment (KnathanM) [inline tests/conftest.py:57]: This fixture isn't used. `test/integration/test_regression_mol.py` uses `mol_data`, while `test/integration/test_regression_rxn.py` uses `rxn_regression_data`. Is there a reason these two integration tests aren't symmetric in that way?

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+def rxn_regression_data(data_dir):
+    df = pd.read_csv(data_dir / "regression/rxn.csv")
+    smis = df["smiles"].to_list()
+    Y = df["ea"].to_numpy().reshape(-1, 1)
+
+    return smis, Y
+
+
+@pytest.fixture
+def mol_mol_regression_data(data_dir):
```
Comment (KnathanM) [inline tests/conftest.py:75]: I don't see `mol_mol_regression_data` and `rxn_mol_regression_data` used. Do you expect that we will have an integration test for them in the future?

File: tests/data/smis.csv
[the remark this answers is not in the document]
Comment (davidegraff) [reply tests/data/smis.csv:1]: arbitrary--I just wanted a simple file containing a list of SMILES strings with no context to make it more clear, i.e., I didn't want to draw the arbitrary SMILES strings from the regression or classification data files. The underlying content might be the same, but the naming of the file and its usage communicates what I want: just a file of SMILES strings.

File: tests/data/test_smiles.csv
Comment (davidegraff) [reply tests/data/test_smiles.csv:1]: no reason to keep both. I just wanted to keep files that were used in at least one place. That is, I didn't want to be the judge of which files were used in certain places.

File: tests/unit/data/test_datapoint.py (hunk 1 above)
Comment (davidegraff) [reply tests/unit/data/test_datapoint.py:46]: I think it's out of scope. I'm not trying to rewrite test logic, just abstract out fixtures where possible

File: tests/conftest.py (hunk 4 above)
Comment (davidegraff) [reply tests/conftest.py:75]: yes, when someones writes integration tests for the `MulticomponentMPNN` class.

File: tests/conftest.py (hunk 3 above)
Comment (davidegraff) [reply tests/conftest.py:57]: I've changed it now and commented the `{mol,rxn}_data` fixtures out for now. Perhaps there will be a use for them in the future.

File: examples/training.ipynb [hunk 5]
```diff
@@ -446,7 +445,19 @@
    "cell_type": "code",
    "execution_count": 15,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
```
[the remark this answers is not in the document]
Comment (kevingreenman) [reply examples/training.ipynb:448]: I agree - I think in general, whenever we make changes to notebooks, the version we publish shouldn't have any errors, and if we're publishing it with the outputs included, the cells should have been run from top to bottom (first cell should have `"execution_count": 1,`, etc.)

File: tests/data/regression_multimolecule.csv
Comment (kevingreenman) [reply tests/data/regression_multimolecule.csv:1]: I see no problem with deleting test files that aren't used anymore. The files I added a couple weeks ago were the result of combining all the test data files that were in the v1 and v2 test folders, so it's fine if some of them aren't used anymore (either at the time of my change, or as a result of this new PR). Shorter test files make more sense to me.

File: tests/data/regression_test_smiles_preds.csv
Comment (kevingreenman) [reply tests/data/regression_test_smiles_preds.csv:1]: I see no problem with deleting test files that aren't used anymore. The files I added a couple weeks ago were the result of combining all the test data files that were in the v1 and v2 test folders, so it's fine if some of them aren't used anymore (either at the time of my change, or as a result of this new PR). Shorter test files make more sense to me.

File: tests/data/reaction_regression.csv
Comment (kevingreenman) [reply tests/data/reaction_regression.csv:1]: I see no problem with deleting test files that aren't used anymore. The files I added a couple weeks ago were the result of combining all the test data files that were in the v1 and v2 test folders, so it's fine if some of them aren't used anymore (either at the time of my change, or as a result of this new PR). Shorter test files make more sense to me.

**Review by kevingreenman — CHANGES_REQUESTED** [review]
let's clear the example notebook outputs and rerun them, and fix the errors shown in pytest before merging

**Comment by KnathanM** [comment]
@kevingreenman Requested changes are addressed.

**Comment by davidegraff** [comment]
why are we testing using our notebooks? these are not actual unit/regression tests, and any behavior that is implicitly tested by them should be covered elsewhere. If anything, the notebooks should go in the docs rather than our testing suite...

**Comment by kevingreenman** [comment]
@davidegraff I think it's a good idea to test the notebooks (added in #562), because we shouldn't trust ourselves to update all of the examples/documentation when we change part of the code (as reported in #559, our notebooks were broken recently, and this could have easily gone unnoticed for a while despite us having good unit/integration tests). This will likely become less of a problem as the v2 codebase becomes more stable (many of the errors in the examples were caused by major refactors of the code), but I don't think it hurts us to continue testing them. I agree that the notebooks should be embedded in the docs - I think that's even more reason to ensure that we're not promoting examples that don't work.

File: tests/data/regression_multimolecule.csv
Comment (kevingreenman) [reply tests/data/regression_multimolecule.csv:1]: It appears this file was in fact used by one of the example notebooks. Could we change the name of the file used by the example notebook to minimize the number of test files we have?

**Review by kevingreenman — APPROVED** [review]
